src/StreamMaster.hpp

```diff
@@ -84,90 +104,177 @@ template <typename Streamer, typename Demux> struct StreamMaster {
     return loop.joinable();
   }
 
-  bool stop() {
-    do_write = false;
-    _stop = true;
-    if (loop.joinable()) {
-      loop.join();
-    }
-    for (auto &d : demux) {
-      this->stop_streamer(streamer[d.topic()]);
-      streamer.erase(d.topic());
+  int stop(void) {
+    try {
+      std::call_once(stop_once_guard,
+                     &FileWriter::StreamMaster<Streamer, Demux>::stop_impl,
+                     this);
+    } catch (std::exception &e) {
+      LOG(0, "Error while stopping: {}", e.what());
     }
-    return !loop.joinable();
+    return !(loop.joinable() || report_thread_.joinable());
   }
 
-  std::map<std::string, typename Streamer::status_type> status() {
-    std::map<std::string, typename Streamer::status_type> stat;
-    for (auto &s : streamer) {
-      stat.emplace(s.first, s.second.status());
+  void report(std::shared_ptr<KafkaW::ProducerTopic> p,
+              const int &delay = 1000) {
+    report_producer_ = p;
+    if (!report_thread_.joinable()) {
+      if (delay < 0) {
+        LOG(2, "Required negative delay in statistics collection: use default");
+        return report(p);
+      }
+      report_thread_ = std::thread(
+          std::bind(&StreamMaster<Streamer, Demux>::report_impl, this,
+                    std::ref(report_producer_), std::ref(delay)));
+    } else {
+      LOG(5, "Status report already started, nothing to do");
     }
-    return stat;
-  }
-
-  rapidjson::Value
-  stats(rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> &a) {
-    return _file_writer_task->stats(a);
-  }
+    return;
+  };
 
   FileWriterTask const &file_writer_task() const { return *_file_writer_task; }
 
+  const StreamMasterError &status() {
+    for (auto &s : streamer) {
+      if (s.second.runstatus().value() < 0) {
+        runstatus = Error::streamer_error;
+      }
+    }
+    return StreamMasterError{runstatus.load()};
```

Comment:
Return ref to temporary, looks like UB to me. Could a simple return by value be an alternative?

Reply:
:white_check_mark: